Trabalho2/grafo.c

```diff
@@ -10,57 +10,412 @@
+    no->nome_user=no_user;
+    no->tamanho = 0;
+    no->ligacoes = NULL;
+
+    return no;
 }
 
 
 no_grafo * no_insere(grafo *g, char *user)
 {
+
+    if(g == NULL){
+        return NULL;
+    }
+
+    if(user == NULL){
+        return NULL;
+    }
+
+    for(int i=0;i<g->tamanho;i++)
+    {
+        if(strcmp(g->nos[i]->nome_user, user) == 0){
+            return NULL;
+        }
+    }
+    
+    no_grafo **aux = g->nos;
+    g->nos = (no_grafo**) realloc(g->nos, sizeof(no_grafo*)*(g->tamanho+1));
+
+    if(g->nos == NULL){
+        g->nos = aux; //Garante que a memoria do g->nos não é perdida mesmo que o realloc não tenha tido sucesso
+        return NULL;
+    }
+    
+    no_grafo* no = cria_no(user);
+
+    if(no == NULL){
+        return NULL;
+    }
+
+    g->nos[g->tamanho] = no;
+    g->tamanho++;
+
+    return no;
+}
+
+ligacao * cria_lig(no_grafo *destino,int peso)
+{
+    if(destino == NULL){
+        return NULL;
+    }
+
+    if(peso < 0){
+        return NULL;
+    }
+
+    ligacao *lig = (ligacao*) malloc(sizeof(ligacao));
     
+    if(lig == NULL){
         return NULL;
-   
+    }
+
+    lig->destino = destino;
+    lig->peso_ligacao = peso;
+
+    return lig;
 }
 
 int  cria_ligacao(no_grafo *origem, no_grafo *destino, int peso)
 {
-   
-    return -1;
+    if(origem == NULL){
+        return -1;
+    }
+
+    if(destino == NULL){
+        return -1;
+    }
+
+    if(peso < 0){
+        return -1;
+    }
+
+    for(int i=0;i<origem->tamanho;i++)
+    {
+        if(strcmp(origem->ligacoes[i]->destino->nome_user, destino->nome_user) == 0){
+            return -1;
+        }
+    }
 
+    ligacao **aux = origem->ligacoes;
+    origem->ligacoes = (ligacao**) realloc(origem->ligacoes, sizeof(ligacao*)*(origem->tamanho+1));
+
+    if(origem->ligacoes == NULL){
+        origem->ligacoes = aux; ////Garante que a memoria do origem->ligacoes não é perdida mesmo que o realloc não tenha tido sucesso
+        return -1;
+    }
+    
+    ligacao *lig = cria_lig(destino, peso);
+
+    if(lig == NULL){
+        return -1;
+    }
+
+    origem->ligacoes[origem->tamanho] = lig;
+    origem->tamanho++;
+
+    return 0;
 }
 
 
 no_grafo * encontra_no(grafo *g, char *nomeU)
 {
+    
+    if(g == NULL){
         return NULL;
+    }
+
+    if(nomeU == NULL){
+        return NULL;
+    }
+
+    for(int i=0;i<g->tamanho;i++)
+    {
+        if(strcmp(g->nos[i]->nome_user, nomeU) == 0){
+            return g->nos[i];
+        }
+    }
+
+    return NULL;
 }
 
 
 
 
 grafo * criaGrafo(tabela_dispersao *td)
 {
-    return NULL;
+    if(td == NULL){
+        return NULL;
+    }
+
+    grafo* g = grafo_novo();
+
+    if(g == NULL){
+        return NULL;
+    }
+
+    elemento *elem;
+    no_grafo *origem, *dest;
+    int msg[2];
+    msg[0] = msg[1] = 0;
+
+    for(int i=0;i<td->tamanho;i++){
+        elem = td->elementos[i];
+        while(elem != NULL)
+        {   
+            origem = no_insere(g, elem->msg->remetente);

```

Comment:
before using elem->msg you should assert that msg is not Null.

Reply:
Indeed, forgot that.